pyinfra/operations/brew.py

```diff
@@ -159,12 +159,13 @@ def casks(
 
 
 @operation
-def tap(src, present=True):
+def tap(src, present=True, url=None):
```

Comment:
We might be able to make `src` optional if `url` is present?

Reply:
Alternative: `src` could accept an URL? `apt.key` does this:

https://github.com/pyinfra-dev/pyinfra/blob/a280e174d85eb41c7851f6d62e6afa89af0c21f4/pyinfra/operations/apt.py#L75-L80

I'm actually not 100% this is the best route (too magic?), would like your opinion!

Reply:
xD I have gotten a bit disillusioned with excess magic since spending time in Golang-land, but the bigger issue with that is that it wouldn't be able to support the full scope of `brew tap [user/repo] [URL]`.

Based on [this search](https://github.com/search?q=%2Fbrew+tap+%5B%5E%5Cs%5D%2B+https.*%2F+language%3AMarkdown&type=code), it seems like the URL argument is usually either

1. `brew tap user/projectname https://github.com/user/projectname` (the behavior we'd support by parsing `src` as a URL and looking at the last two path components)
or
2. `brew tap user/projectname https://github.com/user/homebrew-projectname` (In this case, I'm pretty sure the url is redundant)

However, sometimes people add `.git` at the end of the URL. We could add a special case for that, but there's a less tractable issue:

A (small) percentage of documented uses of `brew tap` don't end with `owner/repo` at all, for example:

```shell
brew tap team23/b5 https://git.team23.de/build/homebrew-b5.git
```

Since our implementation would calculate the tap name as `build/homebrew-b5` (or `build/b5`), it seems like a painful failure mode.

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap('user/project')
brew.tap('https://github.com/user/project')
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap('user/project')
brew.tap('https://github.com/user/project')
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# throw errors?
brew.tap(\"https://git.foo.com/bar.git\") # path is too short
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap('user/project')
brew.tap(\"https://github.com/user/project\")
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# throw errors?
brew.tap(\"https://git.owner.com/foo\") # path is too short
brew.tap(\"https://git.owner.com/foo/bar.git\") 
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\")
brew.tap(\"https://github.com/user/project\")
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# throw errors?
brew.tap(\"https://git.owner.com/foo\") # path is too short
brew.tap(\"https://git.owner.com/foo/bar.git\") 
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\")
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# throw errors?
brew.tap(\"https://git.owner.com/foo\") # path is too short
brew.tap(\"https://git.owner.com/foo/bar.git\") 
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\")
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# throw errors?
brew.tap(\"https://git.owner.com/foo\") # path is too short
brew.tap(\"https://git.owner.com/foo/bar.git\") # path ends in .git
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\")
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# throw errors?
brew.tap(\"https://git.owner.com/foo\") # path is too short
brew.tap(\"https://git.owner.com/foo/bar.git\") # path ends in .git
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\")
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short
brew.tap(\"https://git.owner.com/foo/bar.git\") # path ends in .git
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\")
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short

# unsure
brew.tap(\"https://git.owner.com/foo/bar.git\")
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\")
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short

# unsure
brew.tap(\"https://github.com/owner/project.git\")
brew.tap(\"https://git.owner.com/foo/bar\")
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\")
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short

# unsure
brew.tap(\"https://github.com/owner/project.git\") # should we rstrip(\".git\")?
brew.tap(\"https://git.owner.com/foo/bar\") # should we assume the first path component is the owner?
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", \"https://git.owner.com/foo/bar.git\")

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short

# unsure
brew.tap(\"https://github.com/owner/project.git\") # should we rstrip(\".git\")?
brew.tap(\"https://git.owner.com/foo/bar\") # should we assume the first path component is the owner?
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short

# unsure
brew.tap(\"https://github.com/owner/project.git\") # should we rstrip(\".git\")?
brew.tap(\"https://git.owner.com/foo/bar\") # should we assume the first path component is the owner?
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short

# unsure
brew.tap(\"https://github.com/owner/project.git\") # should we rstrip(\".git\") when calculating?
# could be \"brew tap owner/project https://github.com/owner/project.git\"
# or \"brew tap owner/project.git https://github.com/owner/project.git\"
brew.tap(\"https://git.owner.com/foo/bar\") # should we assume the first path component is the owner?
```

Reply:
I suppose we could \"do both\" -- 

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short

# unsure
brew.tap(\"https://github.com/owner/project.git\")
# could be \"brew tap owner/project https://github.com/owner/project.git\"
# or \"brew tap owner/project.git https://github.com/owner/project.git\"
brew.tap(\"https://git.owner.com/foo/bar\") # should we assume the first path component is the owner?
```

Reply:
I suppose we could \"do both\"

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short

# unsure
brew.tap(\"https://github.com/owner/project.git\")
# could be \"brew tap owner/project https://github.com/owner/project.git\"
# or \"brew tap owner/project.git https://github.com/owner/project.git\"
brew.tap(\"https://git.owner.com/foo/bar\") # should we assume the first path component is the owner?
```

Reply:
I suppose we could \"do both\":

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short

# unsure
brew.tap(\"https://github.com/owner/project.git\")
# could be \"brew tap owner/project https://github.com/owner/project.git\"
# or \"brew tap owner/project.git https://github.com/owner/project.git\"
brew.tap(\"https://git.owner.com/foo/bar\") # should we assume the first path component is the owner?
```

Reply:
I suppose we could \"do both\":

```python
# valid
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")


# unsure
brew.tap(\"https://github.com/owner/project.git\")
# could be \"brew tap owner/project https://github.com/owner/project.git\"
# or \"brew tap owner/project.git https://github.com/owner/project.git\"
brew.tap(\"https://git.owner.com/foo/bar\") # should we assume the first path component is the owner?

# should probably throw an error
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure

```python
brew.tap(\"https://github.com/owner/project.git\")
# could be \"brew tap owner/project https://github.com/owner/project.git\"
# or \"brew tap owner/project.git https://github.com/owner/project.git\"

brew.tap(\"https://git.owner.com/foo/bar\")
```

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure

could be \"brew tap owner/project https://github.com/owner/project.git\"
```python
brew.tap(\"https://github.com/owner/project.git\")
```

```
# or \"brew tap owner/project.git https://github.com/owner/project.git\"

brew.tap(\"https://git.owner.com/foo/bar\")
```

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure

could be `brew tap owner/project https://github.com/owner/project.git`

```python
brew.tap(\"https://github.com/owner/project.git\")
```

```
# or \"brew tap owner/project.git https://github.com/owner/project.git\"

brew.tap(\"https://git.owner.com/foo/bar\")
```

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure


```python
brew.tap(\"https://github.com/owner/project.git\")
```

could be `brew tap owner/project https://github.com/owner/project.git`

```
# or \"brew tap owner/project.git https://github.com/owner/project.git\"

brew.tap(\"https://git.owner.com/foo/bar\")
```

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure


```python
brew.tap(\"https://github.com/owner/project.git\")
```

could be `brew tap owner/project https://github.com/owner/project.git`
or \"brew tap owner/project.git https://github.com/owner/project.git\"
```

brew.tap(\"https://git.owner.com/foo/bar\")
```

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure


```python
brew.tap(\"https://github.com/owner/project.git\")
```

could be `brew tap owner/project https://github.com/owner/project.git`
or \"brew tap owner/project.git https://github.com/owner/project.git\"

```

brew.tap(\"https://git.owner.com/foo/bar\")
```

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure


```python
brew.tap(\"https://github.com/owner/project.git\")
```

could be `brew tap owner/project https://github.com/owner/project.git`
or \"brew tap owner/project.git https://github.com/owner/project.git\"

```
brew.tap(\"https://git.owner.com/foo/bar\")
```

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure


```python
brew.tap(\"https://github.com/owner/project.git\")
```

could be `brew tap owner/project https://github.com/owner/project.git`
or \"brew tap owner/project.git https://github.com/owner/project.git\"

```python
brew.tap(\"https://git.owner.com/foo/bar\")
```

should we assume `foo` is the owner for non-github / bitbucket / ... URLs?

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure


```python
brew.tap(\"https://github.com/owner/project.git\")
```

could be `brew tap owner/project https://github.com/owner/project.git`
or \"brew tap owner/project.git https://github.com/owner/project.git\"

```python
brew.tap(\"https://git.owner.com/foo/bar\")
```

should we assume `foo` is the owner for non-github / bitbucket / ... URLs? It seems clunky to have an allowlist of known git hosts, but

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure


```python
brew.tap(\"https://github.com/owner/project.git\")
```

could be `brew tap owner/project https://github.com/owner/project.git`
or \"brew tap owner/project.git https://github.com/owner/project.git\"

```python
brew.tap(\"https://git.owner.com/foo/bar\")
```

should we assume `foo` is the owner for non-github / bitbucket / ... URLs? It seems clunky to have an allowlist of known git hosts, but otherwise this is still a bit of a footgun

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I suppose we could \"do both\":

## valid

```python
brew.tap(\"user/project\")
brew.tap(\"https://github.com/user/project\") # this feels really nice
brew.tap(\"owner/project\", url=\"https://git.owner.com/foo/bar.git\")
```


## unsure


```python
brew.tap(\"https://github.com/owner/project.git\")
```

could be `brew tap owner/project https://github.com/owner/project.git`
or \"brew tap owner/project.git https://github.com/owner/project.git\"

```python
brew.tap(\"https://git.owner.com/foo/bar\")
```

should we assume `foo` is the owner for non-github / bitbucket / ... URLs? It seems clunky to have an allowlist of known git hosts, but otherwise this is still a bit of a footgun

## invalid

```python
brew.tap(\"https://git.owner.com/foo\") # path is too short
```

Reply:
I think github went wild with some edits there by the looks of it!

I totally agree with the magic thing (funny - also Go that has altered my perspective on such things), let's keep the explicit URL and avoid the magic/complexity.

Reply:
SO I think the only change here is to make the `src` optional!